### PR TITLE
Added learning rate schedulers with warmups

### PR DESCRIPTION
**Pull request by arunppsg**
## Description

I have added learning rate schedulers with warmup in this pull request.
Currently, two warmup schedules are supported:
- linear schedule where the learning rate increases from 0 to `lr` during the warmup period and then
decreases to 0 linearly after the warmup period.
- constant schedule where the learning rate increases from 0 to `lr` during the warmup period and then 
remains constant for rest of the training.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/optimizers.py [hunk 1]
```diff
@@ -550,6 +550,66 @@ def _create_jax_schedule(self):
                                        staircase=self.staircase)
 
 
+class LambdaLRWithWarmup(LearningRateSchedule):
+    """A learning rate scheduler supporting warmup."""
```
Comment (rbharath) [inline deepchem/models/optimizers.py:554]: Can you expand this with a usage example

File: deepchem/models/optimizers.py [hunk 2]
```diff
@@ -550,6 +550,66 @@ def _create_jax_schedule(self):
...
+                during which the learning rate increases linearly between 0 and the initial
+                lr set in the optimizer.
+        """
+        assert warmup_type == 'linear' or 'constant', f'Warmup type {warmup_type} is not supported.'
+        self.initial_rate = initial_rate
+        self.num_warmup_steps = num_warmup_steps
+        self.num_training_steps = num_training_steps
+        self.warmup_type = warmup_type
+
+    def _create_pytorch_schedule(self, optimizer):
```
Comment (rbharath) [inline deepchem/models/optimizers.py:583]: Add some comments/docstrings here. Even if it's private, this is complicated enough that would be useful

**Review by rbharath — COMMENTED** [review]
LGTM once docs are added and CI/lint issues are fixed

File: deepchem/models/optimizers.py (hunk 2 above)
Comment (arunppsg) [reply deepchem/models/optimizers.py:594]: Added docstrings

File: deepchem/models/optimizers.py (hunk 1 above)
Comment (arunppsg) [reply deepchem/models/optimizers.py]: added usage example

**Review by rbharath — APPROVED** [review]
LGTM
